Update version to 0.0.17 and add debug logging functionality to NotificationAPI client SDK

mbasadi opened the pull request:
- Incremented version in package.json and package-lock.json to 0.0.17.
- Introduced a debug mode in the NotificationAPIClientSDK configuration.
- Added debug logging for API calls, WebSocket events, and notification fetching processes to enhance traceability and debugging capabilities.

mbasadi committed:
Update version to 0.0.17 and add debug logging functionality to NotificationAPI client SDK

- Incremented version in package.json and package-lock.json to 0.0.17.
- Introduced a debug mode in the NotificationAPIClientSDK configuration.
- Added debug logging for API calls, WebSocket events, and notification fetching processes to enhance traceability and debugging capabilities.

sahandseifi reviewed (commented):
Comments communicated in slack

mbasadi committed:
Refactor logging in NotificationAPI client SDK to use Logger class

- Introduced a Logger class to handle debug, warn, and error logging.
- Replaced console logging with Logger methods in api and client modules.
- Enhanced logging functionality for API requests, WebSocket events, and notification processes.

sahandseifi commented on lib/client.ts:
```diff
@@ -223,18 +271,37 @@ export const NotificationAPIClientSDK: NotificationAPIClientSDK = {
+      fetchCount++;
+      NotificationAPIClientSDK.logger.log(`Fetch attempt ${fetchCount}`, {
+        oldestReceived,
+        resultCount: result.length
```
log the response, unless it's already logged in the api.ts?

mbasadi replied:
Removed. 
Logged in api.ts

sahandseifi commented on lib/api.ts:
```diff
@@ -7,27 +8,71 @@ export const api = async (
+
+  if (logger) {
+    logger.log('HTTP Request:', {
+      method,
```
Add host

mbasadi replied:
We have the url

sahandseifi commented on lib/api.ts:
```diff
@@ -7,27 +8,71 @@ export const api = async (
+    logger.log('HTTP Request:', {
+      method,
+      url,
+      hasBody: !!data,
```
why do we need hasBody when body is there?

sahandseifi commented on lib/api.ts:
```diff
@@ -7,27 +8,71 @@ export const api = async (
+    const duration = Date.now() - startTime;
+
+    if (logger) {
+      logger.log('HTTP Response:', {
```
Why not log the whole response object?

sahandseifi commented on lib/api.ts:
```diff
@@ -7,27 +8,71 @@ export const api = async (
+        status: res.status,
+        statusText: res.statusText,
+        duration: `${duration}ms`,
+        url
```
Why are we logging url again?

sahandseifi commented on lib/api.ts:
```diff
@@ -7,27 +8,71 @@ export const api = async (
+      return responseData;
+    } catch (e) {
+      if (logger) {
+        logger.warn('Failed to parse response as JSON:', e);
```
Shouldn't this be logger.error?

mbasadi replied:
If the response is empty, we can not parse it. Not necessarily an error. Example: empty notification.
Read this test:
```
  test('should return undefined if response is not JSON', async () => {
    const mockResponse = {
      json: jest.fn().mockRejectedValue(new Error('Invalid JSON'))
    };
    (global.fetch as jest.Mock).mockResolvedValue(mockResponse);

    const result = await api(
      'GET',
      'api.notificationapi.com',
      'testResource',
      'testClient',
      'testUser'
    );

    expect(result).toBeUndefined();
  });
  ```

sahandseifi commented on lib/api.ts:
```diff
@@ -7,27 +8,71 @@ export const api = async (
+  } catch (error) {
+    const duration = Date.now() - startTime;
+
+    if (logger) {
```
Do we need multiple try/catches? What about one that wraps everything in this function?

mbasadi replied:
One try catch was there originally if await res.json(); dose not work to return undefined (a warning log is added)
one try catch is added around everything to log the error

mbasadi committed:
comments

mbasadi committed:
return the try catch

sahandseifi commented on lib/api.ts:
```diff
@@ -7,27 +8,64 @@ export const api = async (
+    });
+
+    if (logger) {
+      logger.log('HTTP Response:', res);
```
Should we remove authorization header from debug?

mbasadi replied:
removed

mbasadi committed:
Enhance API request logging by including headers in the log output

- Added headers to the logging output for API requests to improve traceability.
- Simplified header management by defining headers in a single object.

mbasadi committed:
Refactor API request logging to include headers and improve error reporting

- Removed headers parameter from the API function signature for simplification.
- Enhanced error logging to include headers along with the error details and duration for better traceability.
